### PR TITLE
Configurable execution strategy in `Action`.

### PR DESCRIPTION
**Pull request by andersio**
Implements #291.

`Action`s can now be configured to use different execution strategies other than the default "disable-when-executing" strategy.

~~Two new options have been added to `Action`: `strategy` and `disablesWhenExecuting`.~~

~~The current behaviour is equivalent to `disablesWhenExecuting` being true (the default), regardless of the strategy used.~~

~~So far `FlattenStrategy` and the flattening operator are reused in the implementation.~~ However, the queue coalescer ~~and the buffering cap~~ envisioned in #291 ~~are~~ is not included in the PR, and _might_ require non-trivial changes for maximum reuse.

~~P.S. The `merge` strategy allows `Action` to be concurrent.~~

---

WIP
- [ ] Test cases

### CODE REVIEW

**Comment by mdiep** [comment]
I think it makes more sense to create a custom strategy type for `Action` that encompasses the use of both `strategy` and `disablesWhenExecuting` here. That seems like it'd be the easiest to understand—I don't think all `FlattenStrategy`s will make sense here (what would `.race` do?).

File: Sources/Action.swift [hunk 1]
```diff
@@ -10,15 +10,91 @@ import Result
 /// Actions enforce serial execution. Any attempt to execute an action multiple
 /// times concurrently will return an error.
 public final class Action<Input, Output, Error: Swift.Error> {
+	public enum ExecutionStrategy {
+		/// The default execution strategy. Only one worker is allowed to be
+		/// executing at a time. Any further start attempt during the execution
+		/// is rejected with `ActionError.disabled`.
+		case `default`
```
Comment (mdiep) [inline Sources/Action.swift]: I don't think `default` is a great name for this. We should find a name that describes the behavior.

File: Sources/Action.swift [hunk 2]
```diff
@@ -10,15 +10,91 @@ import Result
...
+		///    been reached; or
+		/// 3. be rejected with `ActionError.disabled` if the capacity limit has
+		///    been reached.
+		///
+		/// - parameters:
+		///   - limit: The concurrent limit of the `Action`.
+		///   - capacity: The capacity limit of the `Action`.
+		///
+		/// - note: `capacity` is inclusive of `limit`.
+		case concurrent(limit: UInt, capacity: UInt)
```
Comment (mdiep) [inline Sources/Action.swift]: I'm not sure that opening it up to concurrency is a good idea. That adds quite a bit of complexity.

Maybe we should limit it to exclusive strategies to start?

File: Sources/Action.swift [hunk 3]
```diff
@@ -10,15 +10,91 @@ import Result
...
 	private let deinitToken: Lifetime.Token
 
 	private let executeClosure: (_ state: Any, _ input: Input) -> SignalProducer<Output, Error>
-	private let eventsObserver: Signal<Event<Output, Error>, NoError>.Observer
 	private let disabledErrorsObserver: Signal<(), NoError>.Observer
 
 	/// The lifetime of the Action.
 	public let lifetime: Lifetime
 
+	private let dispatcher: (SignalProducer<Output, Error>) -> Void
```
Comment (mdiep) [inline Sources/Action.swift]: This could use documentation.

**Comment by andersio** [comment]
Four strategies are left: `concat`, `exclusive`, `latest` and `replay`. That said I am not quite sure about the implementation for now... Meant to use `flatten` but found it a bit of a hack to fit in.

File: Sources/Action.swift [hunk 4]
```diff
@@ -20,19 +20,64 @@ import Result
...
+		init(queueCapacity: UInt, value: Value) {
+			self.queueCapacity = queueCapacity
+			self.value = value
+		}
 	}
 
-	private let execute: (Action<Input, Output, Error>, Input) -> SignalProducer<Output, ActionError<Error>>
-	private let eventsObserver: Signal<Signal<Output, Error>.Event, NoError>.Observer
-	private let disabledErrorsObserver: Signal<(), NoError>.Observer
+	public enum ExecutionStrategy {
```
Comment (mdiep) [inline Sources/Action.swift]: We should make this a `struct` like we did for `FlattenStrategy`.

**Comment by arnopoulos** [comment]
I'm working on a project where we're trying to prevent the user from spamming the server and this would be very helpful. Are there any updates on the progress of this?

**Comment by andersio** [comment]
@arnopoulos I do not have time to move this forward in the foreseeable future. 😞

**Comment by arnopoulos** [comment]
@andersio That's too bad because this definitely could be a useful feature. If I had a deeper understanding of the underlying architecture of ReactiveSwift, I would try to help out

**Comment by ataibarkai** [comment]
is this feature addition still on hold for the foreseeable future?

**Comment by andersio** [comment]
Not being actively worked on.

**Pull request closed without merging**